janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/addedrelations/SimpleAddedRelations.java

```diff
@@ -14,46 +14,109 @@
 
 package org.janusgraph.graphdb.transaction.addedrelations;
 
-import com.google.common.base.Predicate;
 import org.janusgraph.graphdb.internal.InternalRelation;
 
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.List;
+import com.carrotsearch.hppc.LongObjectHashMap;
+import com.carrotsearch.hppc.LongObjectMap;
+import com.carrotsearch.hppc.ObjectArrayList;
+import com.carrotsearch.hppc.cursors.ObjectCursor;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Iterators;
+import javax.annotation.Nonnull;
+
+import java.util.*;
 
 /**
  * @author Matthias Broecheler (me@matthiasb.com)
  */
 
-public class SimpleAddedRelations extends ArrayList<InternalRelation> implements AddedRelationsContainer {
-
+public class SimpleAddedRelations implements AddedRelationsContainer {
```

Comment:
I didn't check the logic yet but I think this class could be simplified (if I am not missing something). What do you think about this implementation?
```
public class SimpleAddedRelations implements AddedRelationsContainer {
    private static final int INITIAL_ADDED_SIZE = 10;
    private final LongObjectMap<ObjectArrayList<InternalRelation>> container;

    public SimpleAddedRelations() {
        this.container = new LongObjectHashMap<>(INITIAL_ADDED_SIZE);
    }

    @Override
    public boolean add(InternalRelation relation) {
        ObjectArrayList<InternalRelation> relations = container.get(relation.longId());
        if (relations == null) {
            relations = new ObjectArrayList<>(1);
            container.put(relation.longId(), relations);
        }
        relations.add(relation);
        return true;
    }

    @Override
    public boolean remove(InternalRelation relation) {
        ObjectArrayList<InternalRelation> relations = container.get(relation.longId());
        if (relations != null) {
            relations.removeFirst(relation);
            if(relations.size()==0){
                container.remove(relation.longId());
            }
        }
        return true;
    }

    @Override
    public Iterable<InternalRelation> getView(Predicate<InternalRelation> filter) {
        return Iterables.filter(this::iterator, filter);
    }

    @Override
    public boolean isEmpty() {
        return container.size()==0;
    }

    @Override
    public Collection<InternalRelation> getAll() {
        return Collections.unmodifiableCollection(new AbstractCollection<InternalRelation>() {
            @Override
            @Nonnull
            public Iterator<InternalRelation> iterator() {
                return SimpleAddedRelations.this.iterator();
            }

            @Override
            public int size() {
                return Iterators.size(SimpleAddedRelations.this.iterator());
            }
        });
    }

    private Iterator<InternalRelation> iterator() {
        return new Iterator<InternalRelation>() {
            private final Iterator<ObjectCursor<ObjectArrayList<InternalRelation>>> iterator = container.values().iterator();
            private Iterator<ObjectCursor<InternalRelation>> child = Collections.emptyIterator();

            @Override
            public boolean hasNext() {
                return child.hasNext() || iterator.hasNext();
            }

            @Override
            public InternalRelation next() {
                if(!child.hasNext()){
                    child = iterator.next().value.iterator();
                }
                return child.next().value;
            }
        };
    }
}
```

Reply:
Also, I didn't check neither the above class nor below, but I think we could optimize memory consumption by using counters. I.e. something like:
```
public class SimpleAddedRelations implements AddedRelationsContainer {
    private static final int INITIAL_ADDED_SIZE = 10;
    private final LongObjectMap<RelationWithCounter> container;

    public SimpleAddedRelations() {
        this.container = new LongObjectHashMap<>(INITIAL_ADDED_SIZE);
    }

    @Override
    public boolean add(InternalRelation relation) {
        RelationWithCounter relations = container.get(relation.longId());
        if (relations == null) {
            container.put(relation.longId(), new RelationWithCounter(relation));
        } else {
            ++relations.amount;
        }
        return true;
    }

    @Override
    public boolean remove(InternalRelation relation) {
        RelationWithCounter relations = container.get(relation.longId());
        if (relations != null) {
            if(--relations.amount <= 0){
                container.remove(relation.longId());
            }
            return true;
        }
        return false;
    }

    @Override
    public Iterable<InternalRelation> getView(Predicate<InternalRelation> filter) {
        return Iterables.filter(this::iterator, filter);
    }

    @Override
    public boolean isEmpty() {
        return container.size()==0;
    }

    @Override
    public Collection<InternalRelation> getAll() {
        return Collections.unmodifiableCollection(new AbstractCollection<InternalRelation>() {
            @Override
            @Nonnull
            public Iterator<InternalRelation> iterator() {
                return SimpleAddedRelations.this.iterator();
            }

            @Override
            public int size() {
                return Iterators.size(SimpleAddedRelations.this.iterator());
            }
        });
    }

    private Iterator<InternalRelation> iterator() {
        return new Iterator<InternalRelation>() {
            private final Iterator<ObjectCursor<RelationWithCounter>> iterator = container.values().iterator();
            private InternalRelation child;
            private int childAmount;

            @Override
            public boolean hasNext() {
                return childAmount > 0 || iterator.hasNext();
            }

            @Override
            public InternalRelation next() {
                if(childAmount == 0){
                    RelationWithCounter relationWithCounter = iterator.next().value;
                    child = relationWithCounter.relation;
                    childAmount = relationWithCounter.amount;
                }
                --childAmount;
                return child;
            }
        };
    }

    private static class RelationWithCounter{
        private final InternalRelation relation;
        private int amount = 1;
        private RelationWithCounter(InternalRelation relation){
            this.relation = relation;
        }
    }
}
```

Reply:
@mad I also thinking, what if we just change the old implementation to `HashSet<InternalRelation>` instead of `ArrayList<InternalRelation>`. Is there any reason why we are using a list for it? If we switch to `Set` implementation then we don't need to build additional logic because same relations will be overridden during addition.